x-pack/plugin/eql/src/main/java/org/elasticsearch/xpack/eql/execution/search/PITAwareQueryClient.java

```diff
@@ -42,46 +48,83 @@ public PITAwareQueryClient(EqlSession eqlSession) {
     protected void search(SearchRequest search, ActionListener<SearchResponse> listener) {
         // no pitId, ask for one
         if (pitId == null) {
-            openPIT(wrap(r -> {
-                pitId = r;
-                searchWithPIT(search, listener);
-            }, listener::onFailure));
-        }
-        else {
+            openPIT(listener, () -> searchWithPIT(search, listener));
+        } else {
             searchWithPIT(search, listener);
         }
     }
 
     private void searchWithPIT(SearchRequest search, ActionListener<SearchResponse> listener) {
         // don't increase the keep alive
         search.source().pointInTimeBuilder(new PointInTimeBuilder(pitId));
+        // get the pid on each response
+        super.search(search, pitListener(SearchResponse::pointInTimeId, listener));
+    }
+
+    @Override
+    protected void search(MultiSearchRequest search, ActionListener<MultiSearchResponse> listener) {
+        // no pitId, ask for one
+        if (pitId == null) {
+            openPIT(listener, () -> searchWithPIT(search, listener));
+        } else {
+            searchWithPIT(search, listener);
+        }
+    }
+
+    private void searchWithPIT(MultiSearchRequest search, ActionListener<MultiSearchResponse> listener) {
+        // don't increase the keep alive
+        for (SearchRequest request : search.requests()) {
+            request.source().pointInTimeBuilder(new PointInTimeBuilder(pitId));
+        }
+
         // get the pid on each request
-        super.search(search, wrap(r -> {
-                pitId = r.pointInTimeId();
+        super.search(search, pitListener(r -> {
+            // get pid
+            for (MultiSearchResponse.Item item : r.getResponses()) {
+                // pick the first non-failing response
+                if (item.isFailure() == false) {
+                    return item.getResponse().pointInTimeId();
+                }
+            }
+            // no results or successful responses, preserve the current pid
+            return pitId;
```

Comment:
@jimczi @dnhatn fyi, this is the current code I've used for updating pit when dealing with multiSearch.